File: frontend/src/services/risk.ts
import axios from "axios";
import { apiUrl } from "./api";

const riskApi = axios.create({
  baseURL: apiUrl,
  withCredentials: true,
  timeout: 60000,
  headers: { "Content-Type": "application/json", "ngrok-skip-browser-warning": "true" },
});

export type RiskLevel = "CRITICAL" | "HIGH" | "MEDIUM" | "LOW";

export type RiskScoreDTO = {
  host_ip: string;
  task_name: string;
  cve_id: string;
  vuln_name: string;
  cvss_score: number;
  epss_score: number;
  epss_percentile: number;
  is_kev: boolean;
  is_ransomware: boolean;
  asset_criticality: string;
  criticality_score: number;
  risk_score: number;
  risk_level: RiskLevel;
};

export type RiskSummaryDTO = {
  total_items: number;
  critical_count: number;
  high_count: number;
  medium_count: number;
  low_count: number;
  top_risks: RiskScoreDTO[];
  last_calculated: string;
};

export type AssetCriticalityDTO = {
  id: number;
  host_ip: string;
  criticality: "crown_jewel" | "high" | "medium" | "low";
  criticality_score: number;
  asset_type: "server" | "database" | "network" | "workstation" | "iot" | "web";
  owner: string;
  business_impact: string;
  created_at: string;
  updated_at: string;
};

export type EPSSStatusDTO = {
  total: number;
  last_sync: string;
  score_date: string;
};

export const GetRiskSummary = async (): Promise<RiskSummaryDTO | null> => {
  try {
    const r = await riskApi.get("/risk/summary");
    return (r.data?.data ?? r.data) as RiskSummaryDTO;
  } catch (e) {
    console.error("GetRiskSummary error:", e);
    return null;
  }
};

export const GetEPSSStatus = async (): Promise<EPSSStatusDTO | null> => {
  try {
    const r = await riskApi.get("/risk/epss/status");
    return r.data as EPSSStatusDTO;
  } catch (e) {
    return null;
  }
};

export const TriggerEPSSSync = async (): Promise<boolean> => {
  try {
    await riskApi.post("/risk/epss/sync");
    return true;
  } catch (e) {
    return false;
  }
};

export const ListAssetCriticality = async (): Promise<AssetCriticalityDTO[]> => {
  try {
    const r = await riskApi.get("/risk/asset-criticality");
    const d = r.data?.data ?? r.data;
    return Array.isArray(d) ? d : [];
  } catch (e) {
    return [];
  }
};

export const CreateAssetCriticality = async (payload: Partial<AssetCriticalityDTO>): Promise<AssetCriticalityDTO | null> => {
  try {
    const r = await riskApi.post("/risk/asset-criticality", payload);
    return r.data?.data ?? null;
  } catch (e) {
    return null;
  }
};

export const UpdateAssetCriticality = async (id: number, payload: Partial<AssetCriticalityDTO>): Promise<boolean> => {
  try {
    await riskApi.patch(`/risk/asset-criticality/${id}`, payload);
    return true;
  } catch (e) {
    return false;
  }
};

export const DeleteAssetCriticality = async (id: number): Promise<boolean> => {
  try {
    await riskApi.delete(`/risk/asset-criticality/${id}`);
    return true;
  } catch (e) {
    return false;
  }
};

File: backend/entity/risk.go
package entity

import (
	"time"

	"gorm.io/gorm"
)

type AppAssetCriticality struct {
	ID               uint           `gorm:"primaryKey;autoIncrement" json:"id"`
	HostIP           string         `gorm:"column:host_ip;uniqueIndex;not null" json:"host_ip"`
	Criticality      string         `gorm:"column:criticality;default:'medium'" json:"criticality"` // crown_jewel, high, medium, low
	CriticalityScore int            `gorm:"column:criticality_score;default:3" json:"criticality_score"` // 1-5
	AssetType        string         `gorm:"column:asset_type;default:'server'" json:"asset_type"` // server, database, network, workstation, iot, web
	Owner            string         `gorm:"column:owner" json:"owner"`
	BusinessImpact   string         `gorm:"column:business_impact;type:text" json:"business_impact"`
	Department       string         `gorm:"column:department" json:"department"`
	OSVersion        string         `gorm:"column:os_version" json:"os_version"`
	EOLDate          *time.Time     `gorm:"column:eol_date" json:"eol_date"` // warranty / end-of-life date, optional
	CreatedAt        time.Time      `json:"created_at"`
	UpdatedAt        time.Time      `json:"updated_at"`
	DeletedAt        gorm.DeletedAt `gorm:"index" json:"-"`
}

type AppEPSSCache struct {
	CVEID        string    `gorm:"primaryKey;column:cve_id" json:"cve_id"`
	EPSSScore    float64   `gorm:"column:epss_score;type:decimal(10,8)" json:"epss_score"`
	Percentile   float64   `gorm:"column:percentile;type:decimal(10,8)" json:"percentile"`
	ModelVersion string    `gorm:"column:model_version" json:"model_version"`
	ScoreDate    string    `gorm:"column:score_date" json:"score_date"`
	FetchedAt    time.Time `gorm:"column:fetched_at" json:"fetched_at"`
	CreatedAt    time.Time `json:"created_at"`
	UpdatedAt    time.Time `json:"updated_at"`
}
